### PR TITLE
Optimize Scrypto Library (Part 4)

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

This PR updates `KeyValueStore` and `ResourceBuilder` to fully support registered types.

It helps reduce WASM code size and runtime transaction cost.

**This change is fully backward-compatible.**

---

Example 1:
```rust
use scrypto::prelude::*;

#[blueprint]
#[types(Hash, Epoch)] // First, register the key type and value type here
mod example {
    struct Example {
        kv_store: KeyValueStore<Hash, Epoch>,
    }

    impl Example {
        pub fn new() -> Global<Example> {
            Self {
                // Then, use `new_with_registered_type` instead of `new` for creating a key value store
                kv_store: KeyValueStore::new_with_registered_type(),
            }
            .instantiate()
            .prepare_to_globalize(OwnerRole::None)
            .globalize()
        }
    }
}
```

---

Example 2:
```rust
use scrypto::prelude::*;

#[derive(ScryptoSbor, NonFungibleData)]
pub struct TestData {
    pub amount_due: Decimal,
}

#[blueprint]
#[types(TestData)] // First, register the non-fungible data type here
mod example {
    struct Example {}

    impl Example {
        pub fn create_non_fungible_resource() {
            // Then, use the `_with_registered_type` interface
            ResourceBuilder::new_ruid_non_fungible_with_registered_type::<TestData>(OwnerRole::None)
                .create_with_no_initial_supply();
        }
    }
}
```

## Testing

New tests added

### For dApp Developers

See the above examples for the new features.

### CODE REVIEW

**Review by talekhinezh — COMMENTED** [review]
Not too big of a fan of the triple generic of KeyValueStore and double generic of NonFungibles. Could we instead require Scrypto to use the same blueprint for types?

**Comment by iamyulong** [comment]
> Not too big of a fan of the triple generic of KeyValueStore and double generic of NonFungibles. 

Tried different approach, but found this to be the less intrusive, fully flexible way of handling additional parameters.

It is also how `HashMap`/`IndexMap` are implemented, when dealing with different `Hasher`.

> Could we instead require Scrypto to use the same blueprint for types?

Personally prefer explicit blueprint over implicit "this blueprint". 

I can potentially autogen wrapper types such as `pub struct RadiswapKeyValueStore<K, V>`, but not sure how much gain we get from this.

**Comment by iamyulong** [comment]
@talekhinezh I've updated the PR to use the new syntax. :-)

File: assets/blueprints/faucet/src/lib.rs [hunk 1]
```diff
@@ -15,7 +16,7 @@ mod faucet {
         ) -> Global<Faucet> {
             Self {
                 vault: Vault::with_bucket(bucket),
-                transactions: KeyValueStore::new(),
+                transactions: KeyValueStore::new_with_registered(),
```
Comment (talekhinezh) [inline assets/blueprints/faucet/src/lib.rs:19]: Minor: not too big of a fan of the name, would like it to be either shorter (e.g. `new_bp_typed`) or more descriptive (`new_with_registered_type`). Middle of the road feels like worse of both worlds?

File: radix-engine-interface/src/api/key_value_store_api.rs [hunk 2]
```diff
@@ -9,6 +9,9 @@ use sbor::rust::prelude::*;
 use sbor::LocalTypeId;
 use sbor::{generate_full_schema, TypeAggregator};
 
+pub const KV_STORE_DATA_SCHEMA_VARIANT_LOCAL: u8 = 0;
+pub const KV_STORE_DATA_SCHEMA_VARIANT_REMOTE: u8 = 1;
```
Comment (talekhinezh) [inline radix-engine-interface/src/api/key_value_store_api.rs:13]: Dead code?

File: scrypto-derive/src/blueprint.rs [hunk 3]
```diff
@@ -222,6 +222,38 @@ pub fn replace_macros(expr: &mut Expr, dependency_exprs: &mut Vec<Expr>) -> Resu
     Ok(())
 }
 
+/// Derive a sensible type identifier from a path, so to pass Radix Engine's constraints.
+pub fn derive_sensible_identifier_from_path(path: &Path) -> Result<String> {
```
Comment (talekhinezh) [inline scrypto-derive/src/blueprint.rs:226]: Perhaps show an example of the transformation in the above documentation just for clarity sake.

File: assets/blueprints/faucet/src/lib.rs (hunk 1 above)
Comment (iamyulong) [reply assets/blueprints/faucet/src/lib.rs:19]: Slightly worried about `_bp_`. Will go `new_with_registered_type`.

File: radix-engine-interface/src/api/key_value_store_api.rs (hunk 2 above)
Comment (iamyulong) [reply radix-engine-interface/src/api/key_value_store_api.rs:13]: `KV_STORE_DATA_SCHEMA_VARIANT_LOCAL` is and the other is not.

It kind of reminds me that I can optimise the local variant as well.

So, both will be used.

File: scrypto-derive/src/blueprint.rs (hunk 3 above)
Comment (talekhinezh) [reply scrypto-derive/src/blueprint.rs:226]: I meant just in the code doc above :)

**Review by talekhinezh — APPROVED** [review]

**Pull request merged**